### DISCUSSION TITLE
Autorot with Magickloaded Tgas

### DISCUSSION THREAD

**Discussion by Moggers** [libvips/libvips#2524] [Q&A] [answered]
Does anyone know if magickloading TGAs and then Autoroting them is meant to work? I'm not sure since just getting the TGA to be opened was a bit of an adventure (had to save the file to disk and then magickload it, as there was no way to hint to vips magickload that the buffer was a TGA, and imagemagick relies on the filename extension to detect its a TGA).

If I use imagemagick directly (ie. `convert --auto-orient`), the image is rotated and then saved as webp, however when using magickload -> autorot -> save, the original orientation of the TGA seems to be ignored (my webps end up upside down).

Some example TGAs:
```
identify -format "%[orientation]" Meropis.tga
BottomLeft%
identify -format "%[orientation]" CapricornCoasts.tga
TopLeft%
```
You can download CapricornCoasts.tga here:
https://blitzserver.net/map/26
And Meropis.tga here:
https://blitzserver.net/map/30

Unfortunately I can't use vipsheader to check this as I can't find a way to force it to delegate to magickload:
```
vipsheader -a CapricornCoasts.tga
vipsheader: VipsForeignLoad: "CapricornCoasts.tga" is not a known file format
```

Installed version is `vips-8.12.0-Thu` if it matters

[My code](https://github.com/blitzserver/blitzserver/blob/master/src/frontend/maps.rs#L417-L423) is using the [rust bindings](https://github.com/olxgroup-oss/libvips-rust-bindings) for vips, so any advice on what the best way to provide a minimal repro is welcome.. Unfortunately I can't use the vips cli as I can find no way to combine `vips magickload` and `vips autorot`.

**Comment by jcupitt** [comment, marked as answer]
Hi @Moggers,

Sorry, the libvips imagemagick loader doesn't support orientation. You're right, it would be a useful feature to add.

You can check this by converting to vips format, since that will keep all the metadata. For example:

```
$ vips magickload CapricornCoasts.tga x.v
$ vipsheader -a x.v
x.v: 3072x1330 uchar, 4 bands, srgb, magickload
width: 3072
height: 1330
bands: 4
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 1
yres: 1
filename: x.v
vips-loader: magickload
magick-date:create: 2021-11-11T09:29:21+00:00
magick-date:modify: 2021-11-11T19:28:44+00:00
n-pages: 1
```

**Reply by jcupitt** [reply]
The imagemagick TGA reader is here:

https://github.com/ImageMagick/ImageMagick/blob/main/coders/tga.c

You can see there's no `IsTGA()` method, so imagemagick can't recognize targa images from the file contents. This is why you had to load via a file.

libvips has it's own recognizer for ICO (which imagemagick also can't recognise), but there's not TGA sniffer. We should probably add one.

**Reply by Moggers** [reply]
Thanks, would you suggest that in my code (until TGA autorot is implemented), I convert to vips format and then call autorot? (will that even work?)

**Reply by jcupitt** [reply]
I don't think that'll work -- the magick loader needs something added to somehow pick up the orientation metadata from libCore.

**Reply by jcupitt** [reply]
... I did a quick PR for the orientation stuff. 

https://github.com/libvips/libvips/pull/2535

**Comment by jcupitt** [comment]
Enhancement issues:

Add a TGA sniffer https://github.com/libvips/libvips/issues/2527

Add orientation support to magickload https://github.com/libvips/libvips/issues/2528
